docs/notebooks/mixed_search_spaces.pct.py

```diff
@@ -0,0 +1,209 @@
+# ---
+# jupyter:
+#   jupytext:
+#     cell_metadata_filter: -all
+#     custom_cell_magics: kql
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.2
+#   kernelspec:
+#     display_name: .venv_310
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Mixed search spaces
+#
+# This notebook demonstrates optimization of mixed search spaces in Trieste.
+#
+# The example function is defined over 2D input space that is a combination of a
+# discrete and a continuous search space. The problem is a modification of the Branin function
+# where one of the input dimensions is discretized.
+
+# %%
+import numpy as np
+import tensorflow as tf
+
+np.random.seed(1793)
+tf.random.set_seed(1793)
+
+# %% [markdown]
+# ## The problem
+#
+# The Branin function is a common optimization benchmark that has three global minima. It is normally
+# defined over a 2D continuous search space.
+#
+# We first show the Branin function over its original continuous search space.
+
+# %%
+from trieste.experimental.plotting import plot_function_plotly
+from trieste.objectives import ScaledBranin
+
+scaled_branin = ScaledBranin.objective
+
+fig = plot_function_plotly(
+    scaled_branin,
+    ScaledBranin.search_space.lower,
+    ScaledBranin.search_space.upper,
+)
+fig.show()
+
+# %% [markdown]
+# In order to convert the Branin function from a continuous to a mixed search space problem, we
+# modify it by discretizing its first input dimension.
+#
+# The discrete dimension is defined by a set of 10 points that are equally spaced, ensuring that
+# the three minimizers are included in this set. The continuous dimension is defined by the
+# interval [0, 1].
+#
+# We observe that the first and third minimizers are equidistant from the middle minimizer, so we
+# choose the discretization points to be equally spaced around the middle minimizer.
+
+# %%
+from trieste.space import Box, DiscreteSearchSpace, TaggedProductSearchSpace
+
+minimizers0 = ScaledBranin.minimizers[:, 0]
+step = (minimizers0[1] - minimizers0[0]) / 4
+points = np.concatenate(
+    [
+        # Equally spaced points to the left of the middle minimizer. Skip the last point as it is
+        # the same as the first point in the next array.
+        np.flip(np.arange(minimizers0[1], 0.0, -step))[:-1],
+        # Equally spaced points to the right of the middle minimizer.
+        np.arange(minimizers0[1], 1.0, step),
+    ]
+)
+discrete_space = DiscreteSearchSpace(points[:, None])
+continuous_space = Box([0.0], [1.0])
+mixed_search_space = TaggedProductSearchSpace(
```

Comment:
here we need to add a comment explaining that we use `TaggedProductSearchSpace`, what a product space is and that we don't give tags